### PR TITLE
Add Action.runAgain extension functions.

### PR DESCRIPTION
**Pull request by Laimiux**
## What
Adding an API to run actions again. This requires to keep the action on the `State` object. 

1. Use case where we always have an action and need to rerun it on some event.
```kotlin
data class State(
  val fetchUserAction: Action<UserData>
)

override fun initialState() = State(
    fetchUserAction = repository.fetchUser()
)

fun evaluate() {
    val onPullToRefreshListener = context.onEvent<Unit> { 
        val newState = state.copy(
            fetchUserAction = state.fetchUserAction.runAgain()
        )
        transition(newState)
    }
}
```

2. Use case where we could potentially have an action (action is nullable)
```kotlin
data class State(
  val updateEmailAction: Action<Status>? = null
)

fun evaluate() {
    val onSaveListener = context.onEvent<Unit> { 
        val newState = state.copy(
            updateEmailAction = RxAction
                .fromObservable { repository.updateEmail(newEmail) }
                .cancelPrevious(state.updateEmailAction)
        )
        transition(newState)
    }
}
```

### CODE REVIEW

**Comment by carrotkite** [comment]
### JaCoCo Code Coverage 80.9% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|
| `com/instacart/formula/internal/ActionWithId` | 100% | 0% | :white_check_mark: |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by mattkranzler5 — APPROVED** [review]
Looks useful!

File: formula/src/test/java/com/instacart/formula/subjects/RunAgainActionFormula.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import com.instacart.formula.Formula
+import com.instacart.formula.Listener
+import com.instacart.formula.Snapshot
+import com.instacart.formula.runAgain
+import com.instacart.formula.subjects.RunAgainActionFormula.Output
+import com.instacart.formula.subjects.RunAgainActionFormula.State
+
+class RunAgainActionFormula : Formula<Unit, State, Output>() {
+    data class State(
+        val action: Action<Unit> = Action.onData(Unit),
```
[the remark this answers is not in the document]
Comment (Laimiux) [reply formula/src/test/java/com/instacart/formula/subjects/RunAgainActionFormula.kt:14]: Currently, we don't store action but instead use an `ActionState<Input>` concept. I'm removing `ActionState` in favor of this simpler approach.

**Comment by deonna** [comment]
Had a question about this while looking at https://github.com/instacart/instacart-android/pull/13466/files -- do we have a way of knowing / enforcing that the `Action` has actually run before calling `runAgain`? Maybe I'm missing it. If we don't, would `runFromFreshState` (that's probably a bad name) be more accurate? Just wondering if calling the extension function `runAgain`  will give consumers the impression that `Formula` enforces that action has already run in that `FormulaContext`.

**Review by deonna — APPROVED** [review]
Left one question.

**Comment by deonna** [comment]
Had a question about this while looking at another PR -- do we have a way of knowing / enforcing that the Action has actually run before calling runAgain? Maybe I'm missing it. If we don't, would runFromFreshState (that's probably a bad name) be more accurate? Just wondering if calling the extension function runAgain will give consumers the impression that Formula enforces that action has already run in that FormulaContext.

**Comment by Laimiux** [comment]
After chatting with a few folks, I've updated the PR to have `runAgain()` and `cancelPrevious` extensions.

**Pull request merged**